cookbooks/fb_storage/libraries/format_devices_provider.rb

```diff
@@ -0,0 +1,638 @@
+          needs_work[:extra_arrays].each do |array|
+            info = node['mdadm'][File.basename(array)]
+            members_set = Set.new(info['members'].map { |x| "/dev/#{x}" })
+            if members_set <= partitions_set
+              Chef::Log.info(
+                'fb_storage: We will rebuild every device in ' +
+                "#{array} **and** it's not an array we're configured to want " +
+                'so we are unmounting and stopping it.',
+              )
+              to_do[:stop_arrays] << array
+            end
+          end
+        end
+
+        to_do
+      end
+
+      # Takes things that need to be converted according to the Storage class
+      # and filters it for what we're allowed to converge
+      def filesystems_to_format(storage)
+        to_do = get_primary_work(storage)
+
+        # all physical hardware, and desired arrays are configured out
+        #
+        # but non-physical hardware requires some extrapolation based on that
+        # data so here we figure out arrays to be filled, cleaned up, etc.
+        fill_in_dynamic_work(to_do, storage)
+      end
+
+      def clear_flag_files(devices, storage)
+        # Any flag files that are there which correspond to disks we configured
+        # or are configured as disks to skip, we remove the flag file.
+        storage.hotswap_disks.each do |disk|
+          if devices.include?(disk) || storage.config[disk]['_skip']
+            filepath = ::File.join(
+              FB::Storage::REPLACED_DISKS_DIR,
+              ::File.basename(disk),
+            )
+            ::File.delete(filepath)
+          end
+        end
+
+        # If we were told by a human to converge all storage, remove that flag
+        # now that we're done.
+        [
+          FB::Storage::CONVERGE_ALL_FILE,
+          FB::Storage::ERASE_ALL_FILE,
+        ].each do |flagfile|
+          if File.exist?(flagfile)
+            ::File.delete(flagfile)
+          end
+        end
+      end
+
+      def set_flag_files
+        format_rules = node['fb_storage']['format'].to_hash
+        indicator_file = FB::Storage::ALREADY_ERASED_ALL_FILE
+        if node.firstboot_tier? && format_rules['firstboot_eraseall'] &&
+            !File.exist?(indicator_file)
+          FileUtils.touch(indicator_file) # ~FB029
+        end
+      end
+
+      # First build missing partitions
+      def partition_storage(devices, storage)
+        devices.each do |device|
+          Chef::Log.debug("fb_storage: Partitioning device #{device}")
+          dev_settings = storage.config[device]
+          unless dev_settings
+            fail "fb_storage: No info for #{device}"
+          end
+
+          # storage handler
+          sh = FB::Storage::Handler.get_handler(device, node)
+          sh.wipe_device
+          sh.prep_device
+          sh.partition_device(dev_settings)
+          sh.condition_device
+        end
+      end
+
+      def device_from_partition(partition)
```

Comment:
this... is the same as `FB::Storage.device_name_from_partition` which is a class method you can call from here, so this method can be dropped and the calls to this in this file should call that.

Reply:
Ah, yes, will refactor this.